scenarios/22_North_Pole.cfg

```diff
@@ -23,6 +23,85 @@
         [/lua]
     [/event]
 
+    #define IMPROVE_STATS_ON_FROZEN UNIT_FILTER
+        #arg DEF
+        60
+        #endarg
+
+        [object]
+            id=frozen_bonus
+            duration=scenario
+            silent=yes
+            take_only_once=no
+            [filter]
+                {UNIT_FILTER}
+            [/filter]
+            [effect]
+                apply_to=defense
+                replace=yes
+                [defense]
+                    frozen={DEF}
+                [/defense]
+            [/effect]
+            [effect]
+                apply_to=movement_costs
+                replace=yes
+                [movement_costs]
+                    frozen=2
+                [/movement_costs]
+            [/effect]
+        [/object]
+    #enddef
+
+    #make summoned ground troops move faster on snow
+    [event]
+        name=unit placed,post advance,post metamorph
+        first_time_only=no
+        [filter]
+            side=1,4,5
+        [/filter]
+
+        {VARIABLE unit_type_when_bonus_set $unit.variables.unit_type_when_bonus_set}
+
+        {IF_VAR unit_type_when_bonus_set not_equals $unit.type (
+            # Skipped if cached.
+            [then]
+                # Cache is cleaned on reloads, so always remove object first to avoid duplication.
+                [remove_object]
+                    id=$unit.id
+                    object_id=frozen_bonus
+                [/remove_object]
+
+                # Note that Fire Elemental doesn't fly.
+                {IF_UNIT_INTERSECT id=$unit.id (
+                    type=EoMa_Carpet_Rider,EoMa_Carpet_Master,EoMa_Dimensional_Gate,EoMa_Dimensional_Gate_II,EoMa_Jinn,EoMa_Great_Jinn,EoMa_Wonderful_Jinn,EoMa_Mystical_Jinn,EoMa_Efreet,EoMa_Great_Efreeti,EoMa_Air_Elemental,EoMa_Air_Avatar,EoMa_Air_God,EoMa_Fire_Avatar,EoMa_Fire_God
+                ) (
+                    # flying
+                ) (
+                    # grounded
+                    {VARIABLE def_bonus 60}
+                    {IF_VAR unit.type equals EoMa_Um (
+                        [then]
+                            {VARIABLE def_bonus 70}
+                        [/then]
+                    )}
+                    {IMPROVE_STATS_ON_FROZEN (id=$unit.id) DEF=$def_bonus}
+                    {CLEAR_VARIABLE def_bonus}
+                )}
+                # Cache unit type ("unit placed" is spammed on [harm_unit], and who knows elsewhen.)
```

Comment:
has_frozen_def_bonus would be more informative than unit_type_when_bonus_set. Unit placed is indeed supposed to happen multiple times for single unit.

Reply:
ideally you have it as part of object filter https://github.com/ProditorMagnus/OrociaRandomMod/blob/9ae304204317ff1299609da22443c695690ad298/lua/helper_functions.lua#L84-L94

Reply:
> has_frozen_def_bonus would be more informative than unit_type_when_bonus_set.

The problem is that some units (e.g. Fire Elemental) must lose its "bonus" (in fact, a replace) when leveling up, so we gotta keep track of the unit type in order to recheck it and reapply the bonus if pertinent.

Reply:
Note that ``has_frozen_def_bonus = unit_type_when_bonus_set ∈ flying_unit_types``. Therefore, the information provided by ``has_frozen_def_bonus`` is also readily available from ``unit_type_when_bonus_set``.

Reply:
You cant add filter inside [effect] for this use case?

Reply:
> You cant add filter inside [effect] for this use case?

I thought ``[filter]`` was a tag for ``[object]`` only. It's great that it works with ``[effect]``. I'd better do that!